<a href="https://colab.research.google.com/github/JSEFERINO/JHSERIES2/blob/main/02_Analisis_Datos2024.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import re
import prince
import warnings
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import OneHotEncoder

plt.style.use("ggplot")
plt.rcParams["figure.figsize"] = (15,6)
warnings.filterwarnings("ignore")
url_base = "https://raw.githubusercontent.com/lacamposm/Metodos_Estadisticos/main/data/"

In [ ]:
# @title Importar datos
datos = pd.read_csv('/content/datos_estudiantes.csv', delimiter=',', encoding='latin-1')
datos.head(5)

CURSO  ASISTENCIA2  ASISTENCIA1  DEFINITIVA  DEFINITIVA2   NRC  \
0  PROBABILIDAD          100           90         3.6         4.30  2314   
1  ESTADISTICAI           70           75         0.9         2.50  1136   
2  PROBABILIDAD           85           95         3.9         3.80  2314   
3  PROBABILIDAD            5            5         2.9         0.50  2314   
4  ESTADISTICAI           20           70         3.7         0.55  1009   

      PROGRAMA  EDAD  PESO  ESTATURA  ...  \
0   F_NEGOCIOS    20    55       160  ...   
1      DERECHO    18    80       185  ...   
2   F_NEGOCIOS    19    60       158  ...   
3     MECANICA    18    72       181  ...   
4  PSICOLOGÍA    19    45       163  ...   

   A: Facilidad para aprender cosas nuevas B: Memoria y atención  \
0                                     BAJO                   ALTO   
1                                     ALTO                   ALTO   
2                                     BAJO                   BAJO   
3                                    MEDIO                   ALTO   
4                                     ALTO                  MEDIO   

  C: Relacionar tu experiencias con lo que aprendes D: Autoestima  \
0                                             MEDIO          ALTO   
1                                              ALTO          ALTO   
2                                              BAJO          BAJO   
3                                              BAJO         MEDIO   
4                                             MEDIO          ALTO   

  E: Actitud hacia el Aprendizaje F: Ambiente Familiar para estudiar  \
0                            ALTO                               BAJO   
1                            ALTO                               ALTO   
2                           MEDIO                              MEDIO   
3                           MEDIO                               ALTO   
4                            ALTO                               BAJO   

  G: Ansiedad académica H: Recursos Educativos  \
0                   BAJO                   BAJO   
1                  MEDIO                  MEDIO   
2                  MEDIO                  MEDIO   
3                  MEDIO                   ALTO   
4                  MEDIO                   BAJO   

  I: Mentalidad para superar adversidades K: Regularidad en el estudio  
0                                    ALTO                         ALTO  
1                                   MEDIO                        MEDIO  
2                                   MEDIO                        MEDIO  
3                                   MEDIO                        MEDIO  
4                                    ALTO                         BAJO  

[5 rows x 26 columns]

In [ ]:
# Verificar el tipo de datos de cada columna
datos.dtypes

CURSO                                                 object
ASISTENCIA2                                            int64
ASISTENCIA1                                            int64
DEFINITIVA                                           float64
DEFINITIVA2                                          float64
NRC                                                    int64
PROGRAMA                                              object
EDAD                                                   int64
PESO                                                   int64
ESTATURA                                               int64
SEXO                                                  object
ESTADO_CIVIL                                          object
ESTRATO                                               object
URBANO                                                object
TRANSPORTE                                            object
GR_SANGUINEO                                          object
 A: Facilidad para aprender cosas nuevas              object
B: Memoria y atención                                object
C: Relacionar tu experiencias con lo que aprendes     object
D: Autoestima                                         object
E: Actitud hacia el Aprendizaje                       object
F: Ambiente Familiar para estudiar                    object
G: Ansiedad académica                                object
H: Recursos Educativos                                object
I: Mentalidad para superar adversidades               object
K: Regularidad en el estudio                          object
dtype: object

In [ ]:
# Definir la función para categorizar los valores
def categorizar_defintiva2(valor):
    if 0 <= valor < 2.5:
        return 'Bajo'
    elif 2.5 <= valor < 3.5:
        return 'Medio'
    elif 3.5 <= valor <= 5:
        return 'Alto'
    else:
        return 'Fuera de rango'  # Para valores que no estén en los rangos especificados

# Aplicar la función a la columna 'DEFINITIVA2' y crear una nueva columna 'Categoría'
datos['Categoria'] = datos['DEFINITIVA2'].apply(categorizar_defintiva2)

# Mostrar los primeros 5 registros para verificar
print(datos.head())

          CURSO  ASISTENCIA2  ASISTENCIA1  DEFINITIVA  DEFINITIVA2   NRC  \
0  PROBABILIDAD          100           90         3.6         4.30  2314   
1  ESTADISTICAI           70           75         0.9         2.50  1136   
2  PROBABILIDAD           85           95         3.9         3.80  2314   
3  PROBABILIDAD            5            5         2.9         0.50  2314   
4  ESTADISTICAI           20           70         3.7         0.55  1009   

      PROGRAMA  EDAD  PESO  ESTATURA  ... D: Autoestima  \
0   F_NEGOCIOS    20    55       160  ...          ALTO   
1      DERECHO    18    80       185  ...          ALTO   
2   F_NEGOCIOS    19    60       158  ...          BAJO   
3     MECANICA    18    72       181  ...         MEDIO   
4  PSICOLOGÍA    19    45       163  ...          ALTO   

  E: Actitud hacia el Aprendizaje F: Ambiente Familiar para estudiar  \
0                            ALTO                               BAJO   
1                            ALTO           

In [ ]:
df2 = datos[['ESTRATO']].join(datos.iloc[:, 16:])

# Mostrar las primeras filas del nuevo DataFrame
print(df2.head())

  ESTRATO  A: Facilidad para aprender cosas nuevas B: Memoria y atención  \
0      II                                     BAJO                   ALTO   
1     III                                     ALTO                   ALTO   
2     III                                     BAJO                   BAJO   
3       V                                    MEDIO                   ALTO   
4      II                                     ALTO                  MEDIO   

  C: Relacionar tu experiencias con lo que aprendes D: Autoestima  \
0                                             MEDIO          ALTO   
1                                              ALTO          ALTO   
2                                              BAJO          BAJO   
3                                              BAJO         MEDIO   
4                                             MEDIO          ALTO   

  E: Actitud hacia el Aprendizaje F: Ambiente Familiar para estudiar  \
0                            ALTO                 

In [ ]:
#remover columnas
df2 = df2.drop(['Categoria_DEFINITIVA2', 'Categoría'], axis='columns')
df2 = df2.astype("category")
df2.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74 entries, 0 to 73
Data columns (total 12 columns):
 #   Column                                             Non-Null Count  Dtype   
---  ------                                             --------------  -----   
 0   ESTRATO                                            74 non-null     category
 1    A: Facilidad para aprender cosas nuevas           74 non-null     category
 2   B: Memoria y atención                             74 non-null     category
 3   C: Relacionar tu experiencias con lo que aprendes  74 non-null     category
 4   D: Autoestima                                      74 non-null     category
 5   E: Actitud hacia el Aprendizaje                    74 non-null     category
 6   F: Ambiente Familiar para estudiar                 74 non-null     category
 7   G: Ansiedad académica                             74 non-null     category
 8   H: Recursos Educativos                             74 non-null     category
 9   I

In [ ]:
# Variables dummies.
X = pd.get_dummies(df2, dtype=int)
X.head()

ESTRATO_I  ESTRATO_II  ESTRATO_III  ESTRATO_IV  ESTRATO_V  \
0          0           1            0           0          0   
1          0           0            1           0          0   
2          0           0            1           0          0   
3          0           0            0           0          1   
4          0           1            0           0          0   

    A: Facilidad para aprender cosas nuevas_ALTO  \
0                                              0   
1                                              1   
2                                              0   
3                                              0   
4                                              1   

    A: Facilidad para aprender cosas nuevas_BAJO  \
0                                              1   
1                                              0   
2                                              1   
3                                              0   
4                                              0   

    A: Facilidad para aprender cosas nuevas_MEDIO  \
0                                               0   
1                                               0   
2                                               0   
3                                               1   
4                                               0   

   B: Memoria y atención_ALTO  B: Memoria y atención_BAJO  ...  \
0                            1                            0  ...   
1                            1                            0  ...   
2                            0                            1  ...   
3                            1                            0  ...   
4                            0                            0  ...   

   H: Recursos Educativos_MEDIO  I: Mentalidad para superar adversidades_ALTO  \
0                             0                                             1   
1                             1                                             0   
2                             1                                             0   
3                             0                                             0   
4                             0                                             1   

   I: Mentalidad para superar adversidades_BAJO  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   

   I: Mentalidad para superar adversidades_MEDIO  \
0                                              0   
1                                              1   
2                                              1   
3                                              1   
4                                              0   

   K: Regularidad en el estudio_ALTO  K: Regularidad en el estudio_BAJO  \
0                                  1                                  0   
1                                  0                                  0   
2                                  0                                  0   
3                                  0                                  0   
4                                  0                                  1   

   K: Regularidad en el estudio_MEDIO  Categoria_Alto  Categoria_Bajo  \
0                                   0               1               0   
1                                   1               0               0   
2                                   1               1               0   
3                                   1               0               1   
4                                   0               0               1   

   Categoria_Medio  
0                0  
1                1  
2                0  
3                0  
4                0  

[5 rows x 38 columns]

In [ ]:
!pip install prince

import prince


In [ ]:
# Realizamos CA en X que es get_dummies del df2
ca = prince.CA(n_components=7, random_state=42).fit(X)
ca

Se extrae la información del análisis

In [ ]:
ca.eigenvalues_summary

eigenvalue % of variance % of variance (cumulative)
component                                                    
0              0.310        14.31%                     14.31%
1              0.215         9.91%                     24.22%
2              0.154         7.10%                     31.32%
3              0.147         6.77%                     38.09%
4              0.132         6.11%                     44.20%
5              0.128         5.89%                     50.09%
6              0.125         5.77%                     55.86%



1.   Se observa que el valor propio más grande es de 0.310 y el mas pequeño es de 0.125
2.   Se observa que los 6 ejees hacen aportes al modelo, obteniendo un 55.86% en la inercia acumulativa.
3. Para interpretar los datos solo se tendrán en cuenta los 2 primeros factores.



In [ ]:
# Tabla con información sobre los 2 primeros factores.
tmp = ca.column_coordinates(X)[[0,1]].rename(columns={0: "factor_1", 1: "factor_2"})
tmp1 = ca.column_contributions_
tmp1 = pd.DataFrame(tmp1).iloc[:, [0,1]]
tmp1.columns = ["Dim1", "Dim2"]
tmp1.index = tmp.index

tmp.merge(tmp1, left_index=True, right_index=True)

factor_1  factor_2  \
ESTRATO_I                                           0.346721 -0.286121   
ESTRATO_II                                          0.067596  0.494627   
ESTRATO_III                                        -0.070620 -0.057528   
ESTRATO_IV                                          0.045373 -0.969961   
ESTRATO_V                                          -1.228180  0.433327   
 A: Facilidad para aprender cosas nuevas_ALTO      -0.706071 -0.525340   
 A: Facilidad para aprender cosas nuevas_BAJO       0.818725 -0.075661   
 A: Facilidad para aprender cosas nuevas_MEDIO     -0.239236  0.481457   
B: Memoria y atención_ALTO                        -1.262261 -0.411243   
B: Memoria y atención_BAJO                         0.808477 -0.100986   
B: Memoria y atención_MEDIO                       -0.251476  0.238288   
C: Relacionar tu experiencias con lo que aprend... -0.935314 -0.285964   
C: Relacionar tu experiencias con lo que aprend...  0.662341  0.034921   
C: Relacionar tu experiencias con lo que aprend...  0.112014  0.158584   
D: Autoestima_ALTO                                 -0.530873 -0.605883   
D: Autoestima_BAJO                                  1.147255 -0.366448   
D: Autoestima_MEDIO                                -0.037410  0.690469   
E: Actitud hacia el Aprendizaje_ALTO               -0.617936 -0.094194   
E: Actitud hacia el Aprendizaje_BAJO                1.508914 -0.576459   
E: Actitud hacia el Aprendizaje_MEDIO               0.711346  0.223642   
F: Ambiente Familiar para estudiar_ALTO            -0.349221 -0.036893   
F: Ambiente Familiar para estudiar_BAJO             1.094445 -0.957208   
F: Ambiente Familiar para estudiar_MEDIO            0.502613  1.340496   
G: Ansiedad académica_ALTO                        -0.202195 -0.025856   
G: Ansiedad académica_BAJO                         0.171338 -0.612175   
G: Ansiedad académica_MEDIO                       -0.063068  0.696474   
H: Recursos Educativos_ALTO                        -1.200060 -0.366053   
H: Recursos Educativos_BAJO                         0.350528 -0.427738   
H: Recursos Educativos_MEDIO                        0.063992  0.571426   
I: Mentalidad para superar adversidades_ALTO       -0.529619 -0.259999   
I: Mentalidad para superar adversidades_BAJO        1.481792 -0.590871   
I: Mentalidad para superar adversidades_MEDIO       0.215559  0.446030   
K: Regularidad en el estudio_ALTO                  -1.043718 -0.490909   
K: Regularidad en el estudio_BAJO                   0.678947 -0.665673   
K: Regularidad en el estudio_MEDIO                 -0.060511  0.747610   
Categoria_Alto                                      0.092540  0.212570   
Categoria_Bajo                                      0.078895 -0.749553   
Categoria_Medio                                    -0.074907  0.083899   

                                                        Dim1      Dim2  
ESTRATO_I                                           0.006547  0.006443  
ESTRATO_II                                          0.000431  0.033373  
ESTRATO_III                                         0.000326  0.000313  
ESTRATO_IV                                          0.000075  0.049360  
ESTRATO_V                                           0.027385  0.004926  
 A: Facilidad para aprender cosas nuevas_ALTO       0.038013  0.030407  
 A: Facilidad para aprender cosas nuevas_BAJO       0.063279  0.000781  
 A: Facilidad para aprender cosas nuevas_MEDIO      0.005611  0.032836  
B: Memoria y atención_ALTO                         0.069421  0.010648  
B: Memoria y atención_BAJO                         0.068825  0.001552  
B: Memoria y atención_MEDIO                        0.007577  0.009831  
C: Relacionar tu experiencias con lo que aprend...  0.063527  0.008581  
C: Relacionar tu experiencias con lo que aprend...  0.036636  0.000147  
C: Relacionar tu experiencias con lo que aprend...  0.001412  0.004090  
D: Autoestima_ALTO                                  0.028652  0.053927

La correlación entre variables y dimensiones principales ayuda a identificar las variables que están más correlacionadas con cada dimensión.


In [ ]:
!pip install altair_viewer
import altair as alt
import altair_viewer

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 844.5/844.5 kB 6.3 MB/s eta 0:00:00


In [ ]:
mca.plot(
    df2,
    x_component=0,
    y_component=1,
    show_column_markers=True,
    show_row_markers=True,
    show_column_labels=False,
    show_row_labels=False
)

alt.LayerChart(...)

Miremos las contribuciones de las variables

In [ ]:
mca.column_contributions_.style.format('{:.0%}')
